### PR TITLE
🎉 (entity selector) sort by name or value

### PR DESCRIPTION
**Pull request by sophiamersmann**
[Cycle 2024.2: Entity selector](https://github.com/owid/owid-grapher/issues/3349) | [Designs](https://www.figma.com/file/X5mOEX8zULS6qyHocUYdmh/Grapher-UI?type=design&node-id=2523%3A6266&mode=design&t=7edFp79OOjz6RENz-1)

## Summary

Adds a dropdown menu to sort by name or indicator, as well as a button to toggle the sort order. Draws a simple bar chart when sorted by an indicator.

## Details

- Indicators considered for sorting are all indicators of the chart that can be formatted numerically (includes `NumberOrStringColumn`) and that are not projections
- Names are sorted asc by default, while numerical values are sorted desc by default
- If sorted by an indicator, then values without data are listed at the end (and are marked with a "No data" label)

## Caveats

- The (display) name of a column that is used as dropdown option is not always nice :( (but we surface it in other places already..)
- We can't draw a bar chart for negative values. For now, we opted to simply not draw anything in that case

## SVG tester

The SVG tester fails due to the changes in #3373 

## Data

- [x] Go through the top 50 charts and make sure all columns have appropriate display names

### CODE REVIEW

**Comment by sophiamersmann** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3447?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3443** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3443?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3447** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3447?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#3436** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3436?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3435** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3435?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3434** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3434?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3373** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3373?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3353** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3353?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#3356](https://github.com/owid/owid-grapher/pull/3356) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3356?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3363](https://github.com/owid/owid-grapher/pull/3363) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3363?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sophiamersmann and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx [hunk 1]
```diff
@@ -99,6 +171,69 @@ export class EntitySelector extends React.Component<{
...
+        const activeSlugs = this.manager.activeColumnSlugs ?? []
+
+        const sortColumns = activeSlugs
+            .map((slug) => this.table.get(slug))
+            .filter(
+                (column) => column.hasNumberFormatting && !column.isProjection
+            )
+
+        return [this.table.entityNameColumn, ...sortColumns]
+    }
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx:195]: Very bad news here, sadly:
This innocent-looking function can take 3+ seconds to run, and is render-blocking.

To repro, go to http://staging-site-entity-selector-sort/grapher/excess-deaths-cumulative-economist-single-entity in a narrow viewport (such that the entity selector doesn't render yet), then resize the window to make the selector show up. The result is that grapher is unresponsive for 6+ seconds, as `LineChart.transformTableForSelection` is running.

I'm a bit confused how this isn't the case on live currently. For some reason, on live, `chartInstance.transformTableForSelection` was called once, and now it's being called 3 times.

I can try to dig into this a bit - for example, to look for perf improvements in `LineChart.transformTableForSelection`...

**Review by marcelgerber — COMMENTED** [review]
Nice work! Sorry for finding the performance issue there, will try to look into it 👀

File: packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx (hunk 1 above)
Comment (marcelgerber) [reply packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx:195]: Fix for `transformTableForSelection` perf at #3486 🌻

Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx:195]: Nice! Thanks a lot for looking into this :)

**Comment by sophiamersmann** [comment]
### Merge activity

* **May 3, 4:04 AM EDT**: @sophiamersmann started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3447).

**Pull request merged**